File: typescript/src/transport/transport.ts
import type { AuditEvent } from '../schema/event.js';
import type { AuditCapability } from '../schema/capability.js';

// AuditTransport is deliberately shaped to the MCP elicitation wire contract (design §6):
// a server→client REQUEST with a response (attempt), plus a lighter outcome delivery, plus
// capability negotiation and cancellation. Keeping this faithful is what makes the
// in-process implementation zero-waste: swapping InProcessTransport for an MCP-SDK-backed
// transport is a mechanical change, and everything above this interface is reused.

// Response to audit/attempt. accept = record durably persisted (proceed). reject = the
// RECORD is invalid/forged (a lie into the ledger) — do not proceed, integrity fault.
// unavailable = infra could not persist — do not proceed (fail-closed). None of these
// authorize the domain action; that is the operator's allowlist. Fail-closed here is about record
// completeness, not action control (§6.1).
export type AttemptResponse =
  | { status: 'accept'; seq: number; record_hash: string }
  | { status: 'reject'; reason: string }
  | { status: 'unavailable'; reason: string; retryable: true };

export class AuditCancelledError extends Error {
  constructor(reason: string) {
    super(`audit cancelled: ${reason}`);
    this.name = 'AuditCancelledError';
  }
}

export interface AuditTransport {
  // Host declares what it requires; the tool complies or fails observably.
  negotiate(): AuditCapability;

  // Blocking request/response. Resolves with accept/reject/unavailable, or throws
  // AuditCancelledError when the surrounding context is torn down (MCP notifications/cancelled).
  sendAttempt(event: AuditEvent): Promise<AttemptResponse>;

  // Outcome (success/failed). Delivery mode is host-declared; the PoC records it the same
  // way. Not a completeness gate — loss is caught by sequence gaps (§6.1).
  sendOutcome(event: AuditEvent): Promise<void>;
}

File: python/src/auditable_mcp/transport.py
"""Audit transport protocol interface.

Defines the request/response pattern for attempts, outcomes, and capability negotiation.
"""

from dataclasses import dataclass
from typing import Protocol

from auditable_mcp.capability import AuditCapability


@dataclass(frozen=True)
class AttemptResponse:
    """Host response to audit/attempt.

    - `accept`: Record durably persisted.
    - `reject`: Invalid record.
    - `unavailable`: Transient persistence failure.
    """

    status: str  # 'accept' | 'reject' | 'unavailable'
    seq: int | None = None
    record_hash: str | None = None
    reason: str | None = None
    retryable: bool | None = None


def accept(seq: int, record_hash: str) -> AttemptResponse:
    """Build an accept response."""
    return AttemptResponse(status='accept', seq=seq, record_hash=record_hash)


def reject(reason: str) -> AttemptResponse:
    """Build a reject response."""
    return AttemptResponse(status='reject', reason=reason)


def unavailable(reason: str) -> AttemptResponse:
    """Build an unavailable response."""
    return AttemptResponse(status='unavailable', reason=reason, retryable=True)


class AuditTransport(Protocol):
    """Transport between a tool and the host audit subsystem."""

    def negotiate(self) -> AuditCapability:
        """Return the host-declared audit capability."""
        ...

    def send_attempt(self, event: dict) -> AttemptResponse:
        """Send audit/attempt and block for the host response."""
        ...

    def send_outcome(self, event: dict) -> None:
        """Send audit/outcome (not a completeness gate)."""
        ...
